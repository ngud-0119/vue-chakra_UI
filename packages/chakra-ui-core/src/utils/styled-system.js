
/**
 * Existential getter function that can be used in any style declaration to get a value
 * from theme, with support for fallback values. This helps prevent errors from
 * throwing when a theme value is missing.
 * @param {Object} obj Theme property
 * @param {String} key Theme key
 * @param {String} def Definition if non-existent
 * @param {*} p
 * @param {*} undef
 */
export const __get = (obj, key, def, p, undef) => {
  key = key && key.split ? key.split('.') : [key]

  for (p = 0; p < key.length; p++) {
    obj = obj ? obj[key[p]] : undef
  }

  return obj === undef ? def : obj
}
