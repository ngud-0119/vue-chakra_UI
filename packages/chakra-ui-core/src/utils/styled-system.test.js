import { __get } from './styled-system'

describe('__get', () => {
  it('should return correct value for given path', () => {
    const theme = {
      baseStyles: {
        Button: {
          bg: 'red.400',
          color: 'black'
        }
      },
      colors: {
        red: {
          400: '#ff0000'
        }
      }
    }

    expect(__get(theme, 1)).toBeUndefined()
    expect(__get(theme, 'someRandomKey')).toBeUndefined()
    expect(__get(theme, 'colors.red.500')).toBeUndefined()
    expect(__get(theme, 'baseStyles.Button.bg._focus')).toBeUndefined()

    expect(__get(theme, 'baseStyles.Button')).toMatchSnapshot()
    expect(__get(theme, 'baseStyles.Button.bg')).toMatchSnapshot()
    expect(__get(theme, 'baseStyles.Button.color')).toMatchSnapshot()
    expect(__get(theme, 'colors.red.400')).toMatchSnapshot()

    expect(__get(theme.baseStyles, 'Button')).toMatchSnapshot()
    expect(__get(theme.baseStyles, 'Button.bg')).toMatchSnapshot()
    expect(__get(theme.baseStyles, 'Button.color')).toMatchSnapshot()
    expect(__get(theme.colors, 'red.400')).toMatchSnapshot()
  })

  it('should return provide default value if specified', () => {
    const theme = {
      baseStyles: {
        Button: {
          bg: 'red.400',
          color: 'black'
        }
      },
      colors: {
        red: {
          400: '#ff0000'
        }
      }
    }

    expect(__get(theme, 1, 'default')).toEqual('default')
    expect(__get(theme, 'someRandomKey', 'default')).toEqual('default')
    expect(__get(theme, 'colors.red.500', 'default')).toEqual('default')
    expect(__get(theme, 'baseStyles.Button.bg._focus', 'default')).toEqual('default')
  })
})
