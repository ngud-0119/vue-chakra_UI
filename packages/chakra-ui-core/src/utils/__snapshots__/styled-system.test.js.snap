// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__get should return correct value for given path 1`] = `
Object {
  "bg": "red.400",
  "color": "black",
}
`;

exports[`__get should return correct value for given path 2`] = `"red.400"`;

exports[`__get should return correct value for given path 3`] = `"black"`;

exports[`__get should return correct value for given path 4`] = `"#ff0000"`;

exports[`__get should return correct value for given path 5`] = `
Object {
  "bg": "red.400",
  "color": "black",
}
`;

exports[`__get should return correct value for given path 6`] = `"red.400"`;

exports[`__get should return correct value for given path 7`] = `"black"`;

exports[`__get should return correct value for given path 8`] = `"#ff0000"`;
